Allow partial shape inference for `tf.nn.conv2d_transpose`

yongtang opened the pull request:
This fix tries to address the issue raised in #8972 where it was not possible to infer a partial shape for
`tf.nn.conv2d_transpose`. In case any dimensions are not defined, the shape will be totally undefined.

This fix utilizes `tensor_util.constant_value_as_shape` so that partial shape is possible.

This fix fixes #8972.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Allow partial shape inference for `tf.nn.conv2d_transpose`

This fix tries to address the issue raised in 8972 where
it was not possible to infer a partial shape for
`tf.nn.conv2d_transpose`. In case any dimensions are not
defined, the shape will be totally undefined.

This fix utilizes `tensor_util.constant_value_as_shape`
so that partial shape is possible.

This fix fixes 8972.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add test case for partial shape inference with `tf.nn.conv2d_transpose`

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

tensorflow-jenkins commented:
Can one of the admins verify this patch?

caisq commented:
@tensorflow-jenkins test this please

cwhipkey commented on tensorflow/python/kernel_tests/conv2d_transpose_test.py:
```diff
@@ -24,8 +24,12 @@
 from tensorflow.python.client import device_lib
```
The C++ shape fn can do this correctly, but has a bug - it is explicitly checking input_tensor instead of just calling MakeShapeFromShapeTensor  (MakeShapeFromShapeTensor used to not be able to handle this case correctly, but now does).

Can you instead change core/ops/nn_ops.cc -> the InputTensorShapeOrUnknown function doesn't need to call input_tensor and check for nullptr, it can just call MakeShapeFromShapeTensor.

yongtang replied:
Thanks @cwhipkey! The `MakeShapeFromShapeTensor` works great. I didn't notice it before. The PR has been updated.

cwhipkey replied:
I think all the callers to InputTensorShapeOrUnknown would benefit from this change - would you mind changing that function instead, to not do the nullptr check and to have the WithRank addition?

yongtang replied:
Thanks @cwhipkey, done.

yongtang committed:
Use MakeShapeFromShapeTensor for shape inference in conv2d_transpose

This commit uses MakeShapeFromShapeTensor for shape inference in
conv2d_transpose instead of using tensor_util.constant_value_as_shape
in python. Based on review comment.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @cwhipkey for the help. The PR has been updated. Please take a look.

yongtang committed:
Convert InputTensorShapeOrUnknown to MakeShapeFromShapeTensor in tf.nn_ops

This fix convert InputTensorShapeOrUnknown to MakeShapeFromShapeTensor in
tf.nn_ops, based on review comments

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @cwhipkey for the review. The PR has been updated. Please take a look.

cwhipkey reviewed (approved):
Can InputTensorShapeOrUnknown be removed now (or, make the changed call sites go back to calling that function, and make it do those 5 lines that are in each shape fn now)?

yongtang committed:
Remove InputTensorShapeOrUnknown

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @cwhipkey. The `InputTensorShapeOrUnknown` and related unit tests have been removed.

sb2nov commented:
Jenkins, test this please.
